File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {

        // binary seacrh

        int n = nums.size();

        int l = 0;
        int r = n-1;

        // binary search algo 
        while(l < r ){

            int mid = l + (r-l)/2; // mid

            // move but which side?

            // consditions

            if(nums[mid] > nums[r]){ // non - sorted part

                l = mid + 1;

            }else{

                r = mid;
            }


        }


        return nums[r];

    }
};

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {

        int n = nums.length;
        int r = n-1;
        int l = 0;

        while(l < r){
            int mid = l + (r-l)/2;

            if(nums[mid] < nums[r]){

                r = mid;

            }else{
                l = mid +1;
            }
        }


        //return nums[l]; both works - breaking condition when r == l -> break
        return nums[r];
    }
}